marimo/_ai/_tools/tools/lint.py

```
@@ -0,0 +1,132 @@
+    diagnostics: list[Diagnostic] = field(default_factory=list)
+
+
+class LintNotebook(ToolBase[LintNotebookArgs, LintNotebookOutput]):
+    """Lint a marimo notebook to check for issues.
+
+    Uses marimo's internal linting engine to check for:
+    - Breaking issues: Problems that prevent the notebook from running
+    - Runtime issues: Problems that may cause unexpected behavior
+    - Formatting issues: Code style and formatting problems
+
+    Args:
+        session_id: The session ID of the notebook to lint.
+
+    Returns:
+        A success result containing all diagnostics found and issue counts.
+    """
+
+    guidelines = ToolGuidelines(
+        when_to_use=[
+            "ALWAYS use this tool after making changes to a marimo notebook to verify you didn't introduce any issues",
+            "ALWAYS use this tool when you want to lint a marimo notebook to check for issues",
+            "When the user asks to check or validate their notebook",
+        ],
+        prerequisites=[
+            "You must have a valid session id from an active notebook",
+        ],
+        avoid_if=[
+            "You just made changes to fix lint issues - explain what you fixed instead of immediately linting again",
+        ],
+        additional_info=(
+            "This tool provides static analysis only and does not execute code. "
+            "Some issues may require running the notebook to fully diagnose."
+        ),
+    )
+
+    async def handle(self, args: LintNotebookArgs) -> LintNotebookOutput:  # type: ignore[override]
+        # Get the session
+        session = self.context.get_session(args.session_id)
+
+        # Get NotebookSerialization IR from session
+        notebook_ir = session.app_file_manager.app.to_ir()
+
+        # Create default rule engine and lint the notebook
+        rule_engine = RuleEngine.create_default()
+        diagnostics = await rule_engine.check_notebook(notebook_ir)
```

maybe wrap in try except

---

got it. It was more of, this function can fail so we can expect it to and return the error. But I'm fine to let it raise nornally.